[HOLD Web-Expensify#36067] Allow joining Restricted Rooms via the Search Page

jasperhuangg opened the pull request:
Relies on https://github.com/Expensify/Web-Expensify/pull/36067

### Fixed Issues
<!---
1. Please replace GH_LINK with a URL link to the GitHub issue this Pull Request is fixing.
2. Please replace PROPOSAL: GH_LINK_ISSUE(COMMENT) with a URL link to your GitHub comment, which contains the approved proposal (i.e. the proposal that was approved by Expensify).

Do NOT add the special GH keywords like `fixed` etc, we have our own process of managing the flow.
It MUST be an entire link to the github issue and your comment proposal ; otherwise, the linking will not work as expected.

Make sure this section looks similar to this (you can link multiple issues using the same formatting, just add a new line):

$ https://github.com/Expensify/App/issues/<number-of-the-issue>
$ https://github.com/Expensify/App/issues/<number-of-the-issue(comment)>

Do NOT only link the issue number like this: $ #<number-of-the-issue>
--->
$ https://github.com/Expensify/Expensify/issues/176852


### Tests/QA Steps
1. Have three accounts, A, B, and C.
    - Ensure A and B are on the same workspace.
    - Ensure C is NOT on the workspace
2. Create a restricted room using account A. 
    - Verify that the room appears in search for account B, but not in the LHN
    - Verify that the room does not appear in search or the LHN for account C.
3. Using account B, select the newly created room from the search page
4. Verify that the Join Room prompt appears.
5. Don't click "Join room" yet, click away to a different report in the LHN. Verify the newly created room disappears from the LHN for account B.
6. Now repeat steps 2-4 again for account B. This time, click "Join room". 
    - Verify the Join Room prompt disappears, and the report action compose component appears
    - In the report details for this room for both accounts A and B, verify the participants are updated to show that both accounts A and B are in the room.
7. Click away to a different report. The newly created restricted room should remain in the LHN for account B.

### Offline tests
1. Have three accounts, A, B, and C.
    - Ensure A and B are on the same workspace.
    - Ensure C is NOT on the workspace
2. Create a restricted room using account A. 
    - Verify that the room appears in search for account B, but not in the LHN
    - Verify that the room does not appear in search or the LHN for account C.
3. Now, using account B, select the newly created room from the search page.
4. Verify that the Join Room prompt appears. Now **_go offline_**.
5. Click "Join room".
    - Verify the Join Room prompt disappears, and the report action compose component appears
    - In the report details for this room for account B, verify the participants are updated to show that both accounts A and B are in the room.
7. Go back online. 
    - In the report details for this room for account A, verify the participants are updated to show that both accounts A and B 

### PR Author Checklist
<!--
This is a checklist for PR authors. Please make sure to complete all tasks and check them off once you do, or else your PR will not be merged!
-->

- [x] I linked the correct issue in the `### Fixed Issues` section above
- [x] I wrote clear testing steps that cover the changes made in this PR
    - [x] I added steps for local testing in the `Tests` section
    - [x] I added steps for the expected offline behavior in the `Offline steps` section
    - [x] I added steps for Staging and/or Production testing in the `QA steps` section
    - [x] I added steps to cover failure scenarios (i.e. verify an input displays the correct error message if the entered data is not correct)
    - [x] I turned off my network connection and tested it while offline to ensure it matches the expected behavior (i.e. verify the default avatar icon is displayed if app is offline)
    - [x] I tested this PR with a [High Traffic account](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#high-traffic-accounts) against the staging or production API to ensure there are no regressions (e.g. long loading states that impact usability).
- [x] I included screenshots or videos for tests on [all platforms](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#make-sure-you-can-test-on-all-platforms)
- [x] I ran the tests on **all platforms** & verified they passed on:
    - [x] Android / native
    - [x] Android / Chrome
    - [ ] iOS / native
    - [x] iOS / Safari
    - [x] MacOS / Chrome / Safari
    - [x] MacOS / Desktop
- [x] I verified there are no console errors (if there's a console error not related to the PR, report it or open an issue for it to be fixed)
- [x] I followed proper code patterns (see [Reviewing the code](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md#reviewing-the-code))
    - [x] I verified that any callback methods that were added or modified are named for what the method does and never what callback they handle (i.e. `toggleReport` and not `onIconClick`)
    - [x] I verified that comments were added to code that is not self explanatory
    - [x] I verified that any new or modified comments were clear, correct English, and explained "why" the code was doing something instead of only explaining "what" the code was doing.
    - [x] I verified any copy / text shown in the product is localized by adding it to `src/languages/*` files and using the [translation method](https://github.com/Expensify/App/blob/4bd99402cebdf4d7394e0d1f260879ea238197eb/src/components/withLocalize.js#L60)
      - [x] If any non-english text was added/modified, I verified the translation was requested/reviewed in #expensify-open-source and it was approved by an internal Expensify engineer. Link to Slack message:
    - [x] I verified all numbers, amounts, dates and phone numbers shown in the product are using the [localization methods](https://github.com/Expensify/App/blob/4bd99402cebdf4d7394e0d1f260879ea238197eb/src/components/withLocalize.js#L60-L68)
    - [x] I verified any copy / text that was added to the app is correct English and approved by marketing by adding the `Waiting for Copy` label for a copy review on the original GH to get the correct copy.
    - [x] I verified proper file naming conventions were followed for any new files or renamed files. All non-platform specific files are named after what they export and are not named "index.js". All platform-specific files are named for the platform the code supports as outlined in the README.
    - [x] I verified the JSDocs style guidelines (in [`STYLE.md`](https://github.com/Expensify/App/blob/main/contributingGuides/STYLE.md#jsdocs)) were followed
- [x] If a new code pattern is added I verified it was agreed to be used by multiple Expensify engineers
- [x] I followed the guidelines as stated in the [Review Guidelines](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md)
- [x] I tested other components that can be impacted by my changes (i.e. if the PR modifies a shared library or component like `Avatar`, I verified the components using `Avatar` are working as expected)
- [x] I verified all code is DRY (the PR doesn't include any logic written more than once, with the exception of tests)
- [x] I verified any variables that can be defined as constants (ie. in CONST.js or at the top of the file that uses the constant) are defined as such
- [x] I verified that if a function's arguments changed that all usages have also been updated correctly
- [x] If a new component is created I verified that:
    - [x] A similar component doesn't exist in the codebase
    - [x] All props are defined accurately and each prop has a `/** comment above it */`
    - [x] The file is named correctly
    - [x] The component has a clear name that is non-ambiguous and the purpose of the component can be inferred from the name alone
    - [x] The only data being stored in the state is data necessary for rendering and nothing else
    - [x] For Class Components, any internal methods passed to components event handlers are bound to `this` properly so there are no scoping issues (i.e. for `onClick={this.submit}` the method `this.submit` should be bound to `this` in the constructor)
    - [x] Any internal methods bound to `this` are necessary to be bound (i.e. avoid `this.submit = this.submit.bind(this);` if `this.submit` is never passed to a component event handler like `onClick`)
    - [x] All JSX used for rendering exists in the render method
    - [x] The component has the minimum amount of code necessary for its purpose, and it is broken down into smaller components in order to separate concerns and functions
- [x] If any new file was added I verified that:
    - [x] The file has a description of what it does and/or why is needed at the top of the file if the code is not self explanatory
- [x] If a new CSS style is added I verified that:
    - [x] A similar style doesn't already exist
    - [x] The style can't be created with an existing [StyleUtils](https://github.com/Expensify/App/blob/main/src/styles/StyleUtils.js) function (i.e. `StyleUtils.getBackgroundAndBorderStyle(themeColors.componentBG`)
- [x] If the PR modifies a generic component, I tested and verified that those changes do not break usages of that component in the rest of the App (i.e. if a shared library or component like `Avatar` is modified, I verified that `Avatar` is working as expected in all cases)
- [x] If the PR modifies a component related to any of the existing Storybook stories, I tested and verified all stories for that component are still working as expected.
- [x] If a new page is added, I verified it's using the `ScrollView` component to make it scrollable when more elements are added to the page.
- [x] I have checked off every checkbox in the PR author checklist, including those that don't apply to this PR.

### Screenshots/Videos
<details>
<summary>Web</summary>

https://user-images.githubusercontent.com/31285285/213806570-4329f3c4-7ba3-4758-9a93-9115344114f9.mov

</details>

<details>
<summary>Mobile Web - Chrome</summary>

https://user-images.githubusercontent.com/31285285/213822644-8805ddaf-1299-4167-9fc8-38c08f80663e.mov

</details>

<details>
<summary>Mobile Web - Safari</summary>

https://user-images.githubusercontent.com/31285285/213821805-d631e65e-550f-46fd-ba20-edeb7ef0bd7b.mp4

</details>

<details>
<summary>Desktop</summary>

https://user-images.githubusercontent.com/31285285/213831054-736e8247-c539-4635-b1e1-fe351e92a9da.mov

</details>

<details>
<summary>iOS</summary>

iOS isn't playing nice with ngrok.. could use some help with this one if possible

</details>

<details>
<summary>Android</summary>

https://user-images.githubusercontent.com/31285285/213830728-6eb86f66-e0a0-413a-bd40-579e795722f1.mov

</details>

jasperhuangg committed:
update report proptypes to include permissions

jasperhuangg committed:
hide restricted rooms from the LHN unless the user has 'read, write' permissions on them

jasperhuangg committed:
update function doc, make comment more descriptive

jasperhuangg committed:
hide report action compose when showing join room prompt (still need to do join room prompt component)

jasperhuangg committed:
wip JoinRoomPrompt component

jasperhuangg committed:
update report details page for restricted rooms

jasperhuangg committed:
wip JoinRoomPrompt component

jasperhuangg committed:
wip JoinRoomPrompt component

jasperhuangg committed:
UI seems to be working as expected, need to handle offline cases/error handling next

jasperhuangg committed:
update optimistic data when creating restricted rooms

shawnborton commented on src/styles/styles.js:
```diff
@@ -2874,6 +2874,37 @@ const styles = {
+    joinRoomPromptContainer: {
+        maxHeight: 92,
+        height: 92,
+        borderRadius: 18,
```
The border radius from Figma is 20. Bonus if we can use a variable for this.

jasperhuangg replied:
Turns out there already is! `variables.componentBorderRadiusRounded`, I've updated it to use this.

shawnborton commented on src/styles/styles.js:
```diff
@@ -2874,6 +2874,37 @@ const styles = {
     },
+
+    joinRoomPromptContainer: {
+        maxHeight: 92,
```
I don't think we need to set a max height here. What is this doing exactly?

jasperhuangg replied:
Ah this was leftover from some debugging I was doing, it should be removed now!

shawnborton commented on src/styles/styles.js:
```diff
@@ -2874,6 +2874,37 @@ const styles = {
+
+    joinRoomPromptContainer: {
+        maxHeight: 92,
+        height: 92,
```
Same comment as above, I don't think we need to set a height.

shawnborton commented on src/styles/styles.js:
```diff
@@ -2874,6 +2874,37 @@ const styles = {
+        height: 92,
+        borderRadius: 18,
+        padding: 20,
+        backgroundColor: themeColors.overlay,
```
I think `themeColors.highlightBG,` is the right color here, not overlay.

shawnborton commented on src/styles/styles.js:
```diff
@@ -2874,6 +2874,37 @@ const styles = {
+    },
+
+    joinRoomPromptTextContainer: {
+        paddingTop: 3.5,
```
What does 3.5 do exactly?

jasperhuangg replied:
When I was in the Figma file I noticed the text had some extra padding (in addition to the 20px given by the outer container) that wasn't being applied to the other elements. But thinking about it now i can probably achieve the same effect with flexbox, will change!

shawnborton commented on src/styles/styles.js:
```diff
@@ -2874,6 +2874,37 @@ const styles = {
+    },
+
+    joinRoomButton: {
+        maxHeight: 40,
```
Why do we need to add a maxHeight? I feel like our standard button style should suffice here, which sits perfectly at 40px (or it should).

shawnborton commented on src/styles/styles.js:
```diff
@@ -2874,6 +2874,37 @@ const styles = {
+
+    joinRoomButton: {
+        maxHeight: 40,
+        margin: 4,
```
What does 4 do here?

Instead, why not add something like `.m4` as a helper class to the actual UI component instead of needing to add in a new style?

shawnborton commented on src/styles/styles.js:
```diff
@@ -2874,6 +2874,37 @@ const styles = {
+    },
+
+    joinRoomButtonText: {
+        fontSize: 13,
```
Why do we need this here? Our default sized buttons should already have this, we shouldn't add special styles for buttons - everything would be styled at the global button level.

jasperhuangg replied:
The default sized buttons have a larger font size (I believe it was 15?).

Are there styles anywhere that let us use size 13 font?

shawnborton replied:
Small buttons that are 28px tall should use 11px font size (micro)

Default buttons that are 40px tall should use 13px font size (label)

Large buttons that are 52px tall should use 15px font size (default)

shawnborton replied:
If it turns out that the default button isn't using the right font size, let's change it globally.

jasperhuangg replied:
Ahhh gotcha, I must have misread the styles when I was originally implementing this, I'll take another look, thanks!

jasperhuangg replied:
Fixed this!

shawnborton commented:
Looks like the mobile/mWeb styles need a lot of love still... for instance, we should either truncate that line below the room name or wrap it to two lines (I prefer truncation, and I think I would make the member count first before the workspace name), and we need to add some margin below the avatar + info so that there is room above the buttons.

jasperhuangg committed:
fix button sizing, remove unused text styling

jasperhuangg committed:
fix text wrap

jasperhuangg committed:
fix background color

jasperhuangg committed:
remove unused style

jasperhuangg committed:
add m1 to spacing, use instead of hard-coded margin

jasperhuangg commented:
@shawnborton I think I fixed the text wrapping on mobile, let me know what you think!

<img width="392" alt="Screenshot 2023-01-23 at 12 59 50 PM" src="https://user-images.githubusercontent.com/31285285/214148977-35796e39-25d8-45cf-8d54-ca1f2cbe0694.png">

Also, I know the `-` separator isn't the same as the midline dot you have in Figma, let me know if you want that instead and I can add it!

I also know it'd be nice to wrap the text at the `-` separator, I think I might be able to get this to work if I use a separate component for the separator, which I'd probably need to do if I addressed the above concern.

jasperhuangg committed:
use correct border radius

jasperhuangg committed:
add failure data

shawnborton commented:
> Also, I know the - separator isn't the same as the midline dot you have in Figma, let me know if you want that instead and I can add it!

Yes, let's add it. Is it easiest to provide it as an svg? It's just a 4px by 4px ellipse. Here it is in full black, which is how we supply all of the icons, and then you can use our existing icon system/styles to add the correct icon color variable to it: [dot-separator.svg.zip](https://github.com/Expensify/App/files/10487761/dot-separator.svg.zip)


> I also know it'd be nice to wrap the text at the - separator, I think I might be able to get this to work if I use a separate component for the separator, which I'd probably need to do if I addressed the above concern.

I think we can just move the members to be at the beginning and then maybe this won't be so bad? Like this: 
<img width="958" alt="image" src="https://user-images.githubusercontent.com/2319350/214247061-7c498b21-6596-4d3a-b733-8428188422ca.png">

jasperhuangg committed:
use dot separator instead of hyphen, put memberCount first

jasperhuangg committed:
fix vertical alignment

jasperhuangg commented:
@shawnborton We're using the svg now! 

I wasn't able to to use flexbox on the outer container surrounding the subtitles to vertically center align the dot separator (for whatever reason applying flex prevents the text from wrapping). Therefore, I did some calculations using the line height and icon size to find the vertical margins needed to center align the dot indicator.

<img width="1004" alt="Screenshot 2023-01-24 at 10 43 07 AM" src="https://user-images.githubusercontent.com/31285285/214380686-48ff47df-3f1f-4148-bee8-d1feb05a8b8a.png">

<img width="387" alt="Screenshot 2023-01-24 at 10 39 51 AM" src="https://user-images.githubusercontent.com/31285285/214380575-82dbdf2c-9452-4f56-aa37-2f8f3e86ce25.png">

jasperhuangg committed:
use read, write, share permissions for participants

jasperhuangg committed:
fix condition

jasperhuangg committed:
update report details page so that the options can re-render if props.report changes

jasperhuangg commented on src/pages/ReportDetailsPage.js:
```diff
@@ -57,43 +56,60 @@ class ReportDetailsPage extends Component {
+        this.renderOptions = this.renderOptions.bind(this);
+    }
+
+    renderOptions() {
```
This was necessary since before, the options were getting defined in the constructor. This worked before because nothing the options relied on could chance.

Now that new users can join the room, the number of participants can change, therefore the Members options needs to be able to re-render to account for this change.

jasperhuangg committed:
remove whitespace

shawnborton commented:
Nice, that looks good to me.

jasperhuangg committed:
Merge branch 'main' of github.com:Expensify/App into jasper-joiningUCPRSearchPage

jasperhuangg committed:
add es locale joinRoom

jasperhuangg committed:
style

jasperhuangg commented:
@rushatgabhane @robertjchen the design for this has been sorted out; this should now be ready for code review, thanks in advance!

jasperhuangg committed:
rename style prop to containerStyles for clarify

jasperhuangg commented:
https://github.com/Expensify/Web-Expensify/pull/36067 was just merged, I'll wait till it's deployed to prod to take the HOLD off this PR!

jasperhuangg committed:
wip fix ios styling

jasperhuangg committed:
fix dot separator display on ios

jasperhuangg commented:
Closing this for the time being via discussion [here](https://expensify.slack.com/archives/CLR7QV2DN/p1675029531686269)
